### PR TITLE
Scaleway object storage driver

### PR DESCRIPTION
**Pull request by reixd**
### Description

This PR adds the capability to connect to Scaleway's Object Storage using libcloud.

### Status

- done, ready for review

### Checklist (tick everything that applies)

- [X] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [X] Documentation
- [X] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [X] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

### CODE REVIEW

**Comment by Kami** [comment]
Thank you for the contribution. Can you please work on getting an ICLA signed - libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes?

File: libcloud/test/storage/test_scaleway.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import unittest
+
+from libcloud.common.types import LibcloudError
+from libcloud.storage.drivers.scaleway import ScalewayStorageDriver
+from libcloud.storage.drivers.scaleway import ScalewayConnectionAWS4
+
+from libcloud.test.secrets import STORAGE_S3_PARAMS
+
+
+class ScalewayStorageDriverTestCase(unittest.TestCase):
```
Comment (Kami) [inline libcloud/test/storage/test_scaleway.py]: Should inherit base s3 tests class here so it verifies that all the base S3 driver related functionality works as expected. IIRC, some existing drivers such as one of the Aurora objects already do that (libcloud/test/storage/test_aurora.py).

File: libcloud/storage/drivers/scaleway.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+        super(ScalewayStorageDriver, self).__init__(key=key,
+                                                 secret=secret,
+                                                 secure=secure,
+                                                 host=host,
+                                                 port=port,
+                                                 url=url)
+
+
+class ScalewayFRParConnection(ScalewayConnectionAWS4):
```
Comment (Kami) [inline libcloud/storage/drivers/scaleway.py]: Single driver with "region" argument for the region should be enough.

We moved away from multiple classes to a single driver class + ``region`` constructor argument quite a while ago. There should be some examples on how to do that in S3 and other drivers.

File: libcloud/storage/drivers/scaleway.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+__all__ = [
+    "ScalewayStorageDriver"
+]
+
+
+class ScalewayConnectionAWS4(SignedAWSConnection, BaseS3Connection):
+    service_name = 's3'
+    version = API_VERSION
+
+    def __init__(self, user_id, key, secure=True, host=None, port=None,
```
Comment (Kami) [inline libcloud/storage/drivers/scaleway.py]: I think it may be enough to inherit from ``S3SignatureV4Connection`` class, no need to duplicate the constructor and override signature version there.

**Comment by Kami** [comment]
Thanks again for the contribution. I added some review comments in-line.

File: libcloud/storage/providers.py [hunk 4]
```diff
@@ -95,6 +95,10 @@
         "DigitalOceanSpacesStorageDriver",
     ),
     Provider.MINIO: ("libcloud.storage.drivers.minio", "MinIOStorageDriver"),
+    Provider.SCALEWAY: ("libcloud.storage.drivers.scaleway", "ScalewayStorageDriver"),
+    Provider.SCALEWAY_FR_PAR: ("libcloud.storage.drivers.scaleway", "ScalewayFRParStorageDriver"),
```
Comment (Kami) [inline libcloud/storage/providers.py]: Same as above, those region specific constants should be removed and the main class should take ``region`` constructor argument.

File: libcloud/test/storage/test_scaleway.py (hunk 1 above)
Comment (reixd) [reply libcloud/test/storage/test_scaleway.py]: Wilco

**Comment by reixd** [comment]
Thank you for the review, will make the changes in the next days.
I will also sign the ICLA.

**Comment by reixd** [comment]
@Kami 
* ICLA signed and sent per mail
* Suggestions worked out and commited

I hope the test "suite" is enough ;)

File: docs/examples/storage/scaleway/upload_example.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+from libcloud.storage.types import Provider
+from libcloud.storage.providers import get_driver
+
+FILE_PATH = "/home/user/myfile.tar.gz"
+
+cls = get_driver(Provider.SCALEWAY)
+driver = cls("api key", "api secret key", host="s3.fr-par.scw.cloud")
```
Comment (Kami) [inline docs/examples/storage/scaleway/upload_example.py]: I believe this example should be updated so a ``region`` instead of ``host`` argument is specified, right?

File: libcloud/storage/drivers/scaleway.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+            url=url,
+            **kwargs,
+        )
+
+    @classmethod
+    def list_regions(self):
+        return REGION_TO_HOST_MAP.keys()
+
+    def get_object_cdn_url(self, *argv):
+        raise LibcloudError(NO_CDN_SUPPORT_ERROR, driver=self)
```
Comment (Kami) [inline libcloud/storage/drivers/scaleway.py]: May want to throw ``NotImplementedError`` instead.

**Comment by reixd** [comment]
@Kami ICLA is through. 
Are we free to go? ;)

**Comment by Kami** [comment]
@reixd Thanks for the reminder - I will have a look as soon as I get a chance :)

File: libcloud/common/aws.py [hunk 7]
```diff
@@ -347,14 +347,15 @@ def _get_credential_scope(self, dt):
         )
 
     def _get_signed_headers(self, headers):
-        return ";".join([k.lower() for k in sorted(headers.keys())])
+        return ";".join([k.lower()
```
Comment (Kami) [inline libcloud/common/aws.py]: I would like to understand reasoning behind this and make sure it doesn't break the generated signature. Can you please provide more context and confirm that?

File: libcloud/storage/drivers/scaleway.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    region_name = "fr-par"
+
+    def __init__(
+        self,
+        key,
+        secret=None,
+        secure=True,
+        host=None,
+        port=None,
+        region=None,
```
Comment (Kami) [inline libcloud/storage/drivers/scaleway.py]: Please also default ``region`` keyword argument to ``fr-par`` here.

File: libcloud/storage/types.py [hunk 9]
```diff
@@ -106,6 +106,10 @@ class Provider(object):
     S3_RGW = "s3_rgw"
     S3_RGW_OUTSCALE = "s3_rgw_outscale"
     MINIO = "minio"
+    SCALEWAY = "scaleway"
+    SCALEWAY_FR_PAR = "scaleway_fr_par"
```
Comment (Kami) [inline libcloud/storage/types.py]: Please remove region specific constants and leave just ``SCALEWAY`` here - as discussed, we moved away from region specific constants and old ones are here for backward compatibility reasons.

**Comment by Kami** [comment]
@reixd Added a couple of more small comments and approved GHA run. Can you please also fix failing lint GHA job?

File: libcloud/common/aws.py (hunk 7 above)
Comment (reixd) [reply libcloud/common/aws.py]: Lets ask @murlock 
Apparently without this change, the S3 Storage in Scaleway does not work.

File: libcloud/storage/drivers/scaleway.py (hunk 8 above)
Comment (reixd) [reply libcloud/storage/drivers/scaleway.py]: good catch.

File: libcloud/storage/types.py (hunk 9 above)
Comment (reixd) [reply libcloud/storage/types.py]: It slipped away...

File: libcloud/common/aws.py (hunk 7 above)
Comment (Kami) [reply libcloud/common/aws.py]: I would hope we have tests for various edge cases for those signature classes, but I would need to dig in.

But yeah, the docs do seem to claim the headers need to be lowecase - https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html.

Would be good to add some test cases which confirm that.

I'll also quickly test this change with S3 when I get a chance, just to confirm this change doesn't break something.

Comment (Kami) [reply libcloud/common/aws.py]: I did a very quick and non extensive test with the S3 driver with those code changes and it seems to work fine.

But yeah, I didn't really dig in much.

Comment (murlock) [reply libcloud/common/aws.py]: Hi everyone
We use only lower case header name:
The data to check signature are setup at https://github.com/openstack-archive/swift3/blob/v1.11/swift3/request.py#L251
and since there is mismatch while preparing data at https://github.com/openstack-archive/swift3/blob/53614c2da837a9a53016b7d3e9911cf96ab39c77/swift3/request.py#L255 a SignatureMismatch is raised

**Comment by reixd** [comment]
> @reixd Added a couple of more small comments and approved GHA run. Can you please also fix failing lint GHA job?

I think I solved the issues. Please start the tests again

**Comment by Kami** [comment]
@reixd Sorry for the delay, I missed your last comment.

LGTM, I will go ahead and merge your change. Thanks again for your contribution!
